File: py/tick_hub/protocols.py
"""Pluggable side-effect boundaries, so the tick engine core stays pure and testable.

The two things a tick does that touch the outside world — running a reminder's shell **gate** and
measuring a health check's **file age** — are expressed as small :class:`typing.Protocol` interfaces
here. The engine takes them as parameters; production uses the default implementations in
:mod:`tick_hub.probes`, and callers can pass deterministic fakes in tests.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Protocol


@dataclass(frozen=True)
class GateResult:
    """The outcome of running a reminder's gate command."""

    #: Exit code of the command (``-1`` when no completed process result exists).
    returncode: int
    #: Captured standard output.
    stdout: str
    #: True iff the command ran to completion (False on launch failure or timeout).
    ok: bool
    #: Human-readable reason when ``ok`` is False; otherwise ``None``.
    error: Optional[str] = None


class GateRunner(Protocol):
    """Runs a reminder's gate command and reports its exit code + stdout."""

    def run(self, cmd: str) -> GateResult:
        """Run ``cmd`` and return its captured result."""
        ...


class FileAgeProbe(Protocol):
    """Reports the age (seconds) of the newest file matching a glob, relative to ``now``.

    Returns ``None`` when nothing matches (so the caller can report ``missing``)."""

    def newest_age_secs(self, pattern: str, now: int) -> Optional[int]:
        """Return the newest matching file's age, or ``None`` when no match is usable."""
        ...

File: rs/tick-hub/src/protocols.rs
//! Pluggable side-effect boundaries used by the tick engine.

/// The outcome of attempting to run a gate command.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct GateResult {
    /// Process exit code, or `-1` when no process result exists.
    pub returncode: i32,
    /// Captured standard output.
    pub stdout: String,
    /// Whether the command actually ran to completion.
    pub ok: bool,
    /// Launch or timeout reason when `ok` is false.
    pub error: Option<String>,
}

impl GateResult {
    /// Construct a completed command result.
    pub fn completed(returncode: i32, stdout: impl Into<String>) -> Self {
        Self {
            returncode,
            stdout: stdout.into(),
            ok: true,
            error: None,
        }
    }

    /// Construct a command-execution failure.
    pub fn failed(error: impl Into<String>) -> Self {
        Self {
            returncode: -1,
            stdout: String::new(),
            ok: false,
            error: Some(error.into()),
        }
    }
}

/// Runs a reminder gate command.
pub trait GateRunner: Sync {
    /// Execute `cmd` and report its exit code and stdout.
    ///
    /// `timeout_secs` overrides the runner's own bound for this one command; `None` keeps the
    /// runner default. A gate that declares a bound and is not held to it would be worse than
    /// one that declares none, so the bound travels with the command rather than being
    /// advisory configuration.
    fn run(&self, cmd: &str, timeout_secs: Option<u64>) -> GateResult;
}

/// Measures the age of the newest file matching a glob.
pub trait FileAgeProbe {
    /// Return seconds since the newest match's mtime, or `None` when missing/unreadable.
    fn newest_age_secs(&self, pattern: &str, now: i64) -> Option<i64>;
}
